client/mysqltest.cc

```diff
@@ -3811,45 +3811,303 @@ void do_remove_files_wildcard(struct st_command *command)
+    } else {
+      /* Extracting the source directory name */
+      if (ds_source->str[strlen(ds_source->str) - 1] == '/') {
+        ds_source->str[strlen(ds_source->str) - 1]= '\0';
+        ds_source->length= ds_source->length - 1;
+      }
+      char *src_dir_name= strrchr(ds_source->str, '/');
+
+      /* Extracting the destination directory name */
+      if (ds_destination->str[strlen(ds_destination->str) - 1] == '/') {
+        ds_destination->str[strlen(ds_destination->str) - 1]= '\0';
+        ds_destination->length= ds_destination->length - 1;
+      }
+      char *dest_dir_name= strrchr(ds_destination->str, '/');
+
+      /*
+        Destination directory might not exist if source directory
+        name and destination directory name are not same.
+        For example, if source is "abc" and destintion is "def",
+        check for the existance of directory "def/abc". If it exists
+        then, copy the files from source directory(i.e "abc") to
+        destination directory(i.e "def/abc"), otherwise create a new
+        directory "abc" under "def" and copy the files from source to
+        destination directory.
+      */
+      if (strcmp(src_dir_name, dest_dir_name)) {
+        dynstr_append(ds_destination, src_dir_name);
+        my_dirend(dest_dir_info);
+        dest_dir_info =
+            my_dir(ds_destination->str, MYF(MY_DONT_SORT | MY_WANT_STAT));
+
+        /* Create destination directory if it doesn't exist */
+        if (!dest_dir_info) {
+          error= my_mkdir(ds_destination->str, 0777, MYF(0)) != 0;
+          if (error) {
+            my_dirend(dest_dir_info);
+            goto end;
+          }
+        }
+      }
+    }
+
+    char dir_separator[2]= {FN_LIBCHAR, 0};
+    dynstr_append(ds_source, dir_separator);
+    dynstr_append(ds_destination, dir_separator);
+
+    /*
+      Storing the length of source and destination
+      directory paths so it can be reused.
+    */
+    size_t source_dir_length= ds_source->length;
+    size_t destination_dir_length= ds_destination->length;
+    ;
+
+    for (uint i= 0; i < src_dir_info->number_of_files; i++) {
+      ds_source->length= source_dir_length;
+      ds_destination->length= destination_dir_length;
+      FILEINFO *file= src_dir_info->dir_entry + i;
+
+      /* Skip the names "." and ".." */
+      if (!strcmp(file->name, ".") || !strcmp(file->name, ".."))
+        continue;
+
+      dynstr_append(ds_source, file->name);
+      dynstr_append(ds_destination, file->name);
+
+      if (MY_S_ISDIR(file->mystat->st_mode))
+        error= (recursive_copy(ds_source, ds_destination) != 0) ? 1 : error;
+      else {
+        DBUG_PRINT("info", ("Copying file: %s to %s", ds_source->str,
+                            ds_destination->str));
+
+        /* MY_HOLD_ORIGINAL_MODES prevents attempts to chown the file */
+        error= (my_copy(ds_source->str, ds_destination->str,
+                         MYF(MY_HOLD_ORIGINAL_MODES)) != 0)
+                    ? 1
```

Comment:
Wouldn't this be more readable as the following? :

```c
if (my_copy(ds_source->str, ds_destination->str,
                         MYF(MY_HOLD_ORIGINAL_MODES)) != 0)
{
  error= 1;
}
```

And probably the one several lines above this too.

Reply:
In fact, it looks like `error` is just a boolean here, should we goto end on error?

Reply:
This is in a for loop and the intention is that the copy will continue to copy other files even if one of the files returns error, so maybe not jump to "end" right away? But I followed your suggestion to rewrite these lines.